Add Dict concept and exercise (#469)
exercism/elm: concepts/dict/about.md

# About

A [`Dict`][dict] in Elm is an immutable dictionary of zero or more key-value pairs.

> Type annotations for dicts are be written as follows

Reviewer: ```suggestion
Type annotations for dicts are written as follows
```

```elm
Dict String (List Int) --> a dict with String keys and List Int values
Dict Int String --> a dict with Int keys and String values
```

Dicts can be created as follows:

```elm
empty : Dict Int String
empty = Dict.empty

singleValue = Dict.singleton 5 "Value for key 5" --> Dict Int String

twoValues = Dict.fromList [ ( "Alice", 0 ), ( "Bob", 1 ) ] --> Dict Int String
```

Items can be retrieved using `get`.
As a key may or may not be present in the dict, the result will be wrapped in a `Maybe`.

```elm
alice = Dict.fromList [ ( "Alice", 0 ) ]

> keyExists = Dict.get 0 alice--> Just "Alice"

Reviewer 2: ```suggestion
keyExists = Dict.get 0 alice --> Just "Alice"
```

keyNotPresent = Dict.get 1 alice --> Nothing
```

Items can be added using `insert`.
Items for keys that are already in the dict get replaced.

```elm
alice = Dict.fromList [ ( "Alice", 0 ) ]
aliceAndBob = Dict.insert 1 "Bob" alice  --> Dict.fromList [ ( "Alice", 0 ), ( "Bob", 1 ) ]
```

Items can be updated using `update`.

```elm

> alice = Dict.fromList [ ( "Alice", 0 ) ]
> aliceUpperCase = Dict.update 0 (Maybe.map toUpper) alice --> Dict.fromList [ ( "ALICE", 0 ) ]

Reviewer 2: ```suggestion
alice = Dict.fromList [ ( 0, "Alice" ) ]
aliceUpperCase = Dict.update 0 (Maybe.map String.toUpper) alice --> Dict.fromList [ ( 0, "ALICE" ) ]
```

```

Items can be removed using `remove`.

```elm
alice = Dict.fromList [ ( "Alice", 0 ) ]
empty = Dict.remove 0 alice --> Dict.empty
```

Multiple items can be removed using `filter`.

```elm
aliceAndBob = Dict.fromList [ ( "Alice", 1 ), ( "Bob", 0 ) ]
bob = Dict.filter (\playerName goalCount) -> playerName == "Bob") aliceAndBob --> Dict.fromList [ ( "Bob", 0 ) ]
```

Items can be transformed using `map`.

```elm
alice = Dict.fromList [ ( "Alice", 0 ) ]
empty = Dict.map (\playerName goalCount -> goalCount + 1) --> Dict.fromList [ ( "Alice", 1 ) ]
```

Dicts can be combined / transformed using `merge`.

```elm
aliceAndBob = Dict.fromList [ ( "Alice", 1 ), ( "Bob", 1 ) ]
bobAndCedd = Dict.fromList [ ( "Bob", 1 ), ( "Cedd", 1 ) ]

> empty = Dict.merge
>         -- when only in aliceAndBob
>         (\playerName goalCount mergedGoalCounts -> Dict.insert playerName goalCount mergedGoalCounts)
>         -- when in aliceAndBob and bobAndCedd
>         (\playerName aliceAndBobGoalCount bobAndCeddGoalCount mergedGoalCounts -> Dict.insert playerName (aliceAndBobGoalCount + bobAndCeddGoalCount) mergedGoalCounts)
>         -- when only in bobAndCedd
>         (\playerName goalCount mergedGoalCounts -> Dict.insert playerName goalCount mergedGoalCounts)
>         -- the two dicts to merge
>         aliceAndBob
>         bobAndCedd
>         -- the initial state of the merge
>         Dict.empty
> --> Dict.fromList [ ( "Alice", 1 ), ( "Bob", 2 ), ( "Cedd", 1 ) ]

Reviewer 2: I would propose some changes here because students could copy paste the example to solve the exercise. I also made the variable names shorter for readability (and to examples above for consistency).

```suggestion
firstDictCountsDouble = Dict.merge
        -- when only in aliceAndBob
        (\name count mergedCounts -> Dict.insert name (2 * count) mergedCounts)
        -- when in aliceAndBob and bobAndCedd
        (\name aliceAndBobCount bobAndCeddCount mergedCounts -> Dict.insert name (2 * aliceAndBobCount + bobAndCeddCount) mergedCounts)
        -- when only in bobAndCedd
        (\name count mergedCounts -> Dict.insert name count mergedCounts)
        -- the two dicts to merge
        aliceAndBob
        bobAndCedd
        -- the initial state of the merge
        Dict.empty
--> Dict.fromList [ ( "Alice", 2 ), ( "Bob", 3 ), ( "Cedd", 1 ) ]
```
